### PR TITLE
[ASA-252][AK,RV] Overseas Agents - Create supervisory body details page

### PR DESCRIPTION
**Pull request by rvijayan**

### CODE REVIEW

File: app/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/ApplicationController.scala [hunk 1]
```diff
@@ -1,23 +1,49 @@
...
   val authConnector: DefaultAuthConnector,
   val env: Environment,
-  isValidApplicant: AgentAffinityNoHmrcAsAgentAuthAction)(implicit val configuration: Configuration)
+  validApplicantAction: AgentAffinityNoHmrcAsAgentAuthAction,
+  sessionStoreService: SessionStoreService)(implicit val configuration: Configuration, ec: ExecutionContext)
     extends FrontendController with I18nSupport {
 
-  def root: Action[AnyContent] = isValidApplicant.async { implicit request =>
-    ???
+  def root: Action[AnyContent] = validApplicantAction.async { implicit request =>
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/ApplicationController.scala]: lets move to StartController

File: app/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/auth/BasicAuthAction.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  implicit ec: ExecutionContext)
+    extends BasicAuthAction with AuthorisedFunctions with AuthRedirects {
+
+  def invokeBlock[A](request: Request[A], block: (Request[A]) => Future[Result]): Future[Result] = {
+    implicit val hc = HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
+
+    authorised(AuthProviders(GovernmentGateway)) { block(request) }
+      .recover {
+        case _: NoActiveSession =>
+          val isDevEnv =
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/auth/BasicAuthAction.scala]: lets move the isDev logic into FrontendModule, and just have a val for it true or false

File: app/uk/gov/hmrc/agentoverseasapplicationfrontend/models/AgentApplication.scala [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.hmrc.agentoverseasapplicationfrontend.models
+
+import play.api.libs.json.Json
+import uk.gov.hmrc.agentmtdidentifiers.model.Utr
+import uk.gov.hmrc.domain.Nino
+
+case class AgentApplication(amlsDetails: Option[AmlsDetails] = None)
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentoverseasapplicationfrontend/models/AgentApplication.scala]: amls details will always be mandatory

File: conf/messages [hunk 4]
```diff
@@ -8,6 +8,27 @@ error.maxLength=You can only enter up to {0} characters
 error.email=Enter a valid email address
 
 # Service specific messages
-app.name=agent-overseas-application
-servicename.titleSuffix=agent-overseas-application - GOV.UK
+app.name=agent-services - apply-from-outside-uk
+servicename.titleSuffix=agent-services - apply-from-outside-uk - GOV.UK
```
Comment (arkowalski) [inline conf/messages]: update to reflect prototype

File: it/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/ApplicationControllerISpec.scala [hunk 5]
```diff
@@ -1,11 +1,49 @@
...
+        "amls.hint.expandable",
+        "amls.hint.expandable.p1"
+      )
+
+      result should containSubstrings("https://www.gov.uk/guidance/get-an-hmrc-agent-services-account",
+        routes.SignOutController.signOut().url)
+    }
+  }
+
+  "POST /money-laundering" should {
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/ApplicationControllerISpec.scala]: need a test for bad details submitted, to know valid errors are shown

File: it/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/ApplicationControllerISpec.scala [hunk 6]
```diff
@@ -1,11 +1,49 @@
...
+    "redirect to contact-details form" in {
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/ApplicationControllerISpec.scala]: remove the "form"

File: app/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/auth/BasicAuthAction.scala [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            if (env.mode.equals(Mode.Test)) false else config.getString("run.mode").forall(Mode.Dev.toString.equals)
+          toGGLogin(if (isDevEnv) s"http://${request.host}${request.uri}" else s"${request.uri}")
+      }
+  }
+}
+
+@ImplementedBy(classOf[BasicAuthActionImpl])
+trait BasicAuthAction extends ActionBuilder[Request]
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/auth/BasicAuthAction.scala]: this is not specific to this PR/service, but i am not happy with the `ActionBuilder ` approach  
what if we have  to write more than one auth method each retrieves different credentials or  authProviders.. ? then do we have to write one separate `***AuthAction `class for each type of auth method ? 
the AuthAction that we have in `agent-subscription` backend looks better (it has two auth methods )

Comment (rvijayan) [reply app/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/auth/BasicAuthAction.scala]: Yes, we will end up creating different auth actions just like we create different auth methods if we need to retrieve different credentials or authProviders. At present, we didn't have scenarios though. The best part with creating auth actions is that we can use it for different controller action behaviours. eg: AuthAction.async {}, AuthAction {}, AuthAction.async(parse.json) {}. 
This is not possible with auth methods. Also play framework encourages us to create such custom actions.

File: it/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/StartControllerISpec.scala [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+  "GET /not-agent" should {
+    "display the non-agent  page when the current user is logged in" in {
+
+      val result = await(controller.showNotAgent(basicRequest(FakeRequest())))
+
+      status(result) shouldBe 200
+    }
+
+    "redirect to the company-auth-frontend sign-in page when the current user is not logged in" in {
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/StartControllerISpec.scala]: this should be captured in the basicAuth tests

**Pull request merged**